### PR TITLE
also transform query param

### LINKED ISSUE #6: Transformers ignore queries
Queries return a `Key` type which is a `[]byte` but can't directly be cast to string.

### PR DESCRIPTION
**Pull request by woutslakhorst**
Fixes #6

### CODE REVIEW

File: index.go [hunk 1]
```diff
@@ -73,6 +73,8 @@ type IndexPart interface {
 	// Transform is a function that alters the value to be indexed as well as any search criteria.
 	// For example LowerCase is a Transform function that transforms the value to lower case.
 	Transform(value interface{}) interface{}
+	// Transformer return the Transform function
+	Transformer() Transform
```
Comment (reinkrul) [inline index.go:77]: This feels really odd, having dependency which is both invoked (using `Transform`) but can be returned as well (`Transformer`)?

File: json.go [hunk 2]
```diff
@@ -96,9 +96,17 @@ func (j jsonIndexPart) Transform(value interface{}) interface{} {
 	if s, ok := value.(string); ok {
```
Comment (reinkrul) [inline json.go:96]: You could do this more esthetically using a type-select;

```golang
	switch s := value.(type) {
	case string:
		return j.transformer(s)
	case Key:
		return Key(j.transformer(s.String()))
	default:
		return value
	}
```

File: search.go [hunk 3]
```diff
@@ -40,7 +41,7 @@ type QueryPart interface {
 	Seek() (Key, error)
 
 	// Condition returns true if given key falls within this condition.
```
Comment (reinkrul) [inline search.go:43]: can you add some doc on the transform arg?

File: search.go [hunk 4]
```diff
@@ -145,7 +153,14 @@ func (p prefixPart) Seek() (Key, error) {
 	return toBytes(p.value)
 }
 
-func (p prefixPart) Condition(key Key) (bool, error) {
+func (p prefixPart) Condition(key Key, transform Transform) (bool, error) {
+	if transform != nil {
+		s, ok := p.value.(string)
```
Comment (reinkrul) [inline search.go:158]: only strings are transformed?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: search.go (hunk 4 above)
Comment (woutslakhorst) [reply search.go]: changed to `bytes.HasPrefix`

File: index.go (hunk 1 above)
Comment (woutslakhorst) [reply index.go]: changed fn definition to use interface{}

**Review by reinkrul — APPROVED** [review]

**Pull request merged**